client/src/utils/Store.js

```diff
@@ -0,0 +1,126 @@
+    this.plot = '';
+    this.db = new IDB();
+    this.init();
+  }
+
+  init = async () => {
+    const [models, keys] = await Promise.all([
+      this.db.getAll(),
+      this.db.keys(),
+    ]);
+    this.models = models;
+    this.keys = keys;
+  }
+
+  /**
+   * Gets plot from server
+   * @param {String} fig_type, type of figure to be requested from server
+   * @return {String} image data base64 string, or null if request fails
+   */
+  createPlot = async (fig_type) => {
+    // may need to call createModel on all items before getting plot
+    try {
+      const { data } = await axios.post(`${baseurl}/plot`, {
+        fig_type,
+        img_type: 'png',
+      });
+      this.plot = `data:image/png;base64,${data.figure}`;
+    } catch (error) {
+      console.error(error);
+      // better error messaging here
+    }
+  }
+
+  /**
+   * Sends model data to server to create Tracer Halo Model Object
+   * Also saves model into indexed db
+   * @param {Object} model model data
+   * @param {String} name model name
+   */
+  createModel = async (model, name) => {
+    const params = {};
+    Object.values(model).forEach((value) => {
+      Object.assign(params, value);
+    });
+    try {
+      await axios.post(`${baseurl}/create`, {
+        params,
+        label: name,
+      });
+      await Promise.all([this.setModel(model, name), this.createPlot()]);
+    } catch (error) {
+      console.error(error);
+      // better error messaging here
+    }
+  }
+
+  /**
+   * Updates a model
+   * @param {Object} model model to update
+   * @param {String} label label to update model
+   */
+  updateModel = async (model, name) => {
+    try {
+      /* some api call here */
+      await Promise.all([this.setModel(model, name), this.createPlot()]);
+    } catch (error) {
+      console.error(error);
+      // better error handling here, some vue event?
+    }
+  }
+
+  /**
+   *
+   * @param {*} model
+   * @param {*} name
+   */
+  cloneModel = async (model, name) => {
+    try {
+      // unknown api call somewhere,
+      // name may be old name for api call
+      // use getModel to create the cloned model
+      await Promise.all([this.setModel(model, name), this.createPlot()]);
+    } catch (error) {
+      console.error(error);
+    }
+  }
+
+  /**
+   * Gets (clones) a model at label, keeps function pure.
+   * @param {String} label
+   * @returns {Object} A copy of the target model, or null
+   */
+  getModel = (name) => clonedeep(this.models[name])
+
+  /**
+   * gets all keys
+   * @returns {Array<String>} keys
+   */
+  getKeys = () => this.keys
+
+  /**
+   * Sets a model at label
+   * @param {Object} model the model object to set
+   * @param {String} name the name to set the model
+   */
+  setModel = async (model, name) => {
+    // puts model inside idb after successful api call
+    await this.db.put(name, {
+      name,
+      model,
+    });
+    this.models[name] = model;
+    this.keys = this.db.keys();
```

Comment:
Should probably grab keys from this.models instead of this.db (faster since its a js object)

Reply:
In fact now that I'm thinking about it it may better to maintain the specific array @ this.keys and mutating it instead of overriding it with a reference to a new array whenever it needs to be adjusted. That lets us keep a reference to that object in home and prevents us from needing a call to keys whenever we do anything.

adding to an array is just keys.push()
removing from it would require a helper function
```
function removeIitem(arr, value) {
  var index = arr.indexOf(value);
  if (index > -1) {
    arr.splice(index, 1);
  }
  return arr;
}
```

Create/clone would add to it.
Update wouldn't do anything with it.
Delete would remove from it.

Reply:
You're doing two for loops to get that, might as well just get it from the db. The keys() in the async call is constant time.

Reply:
it's already in the models, I save the names inside the models object.  The keys just gets an array.